### PR TITLE
CSV Rejects table

### PR DESCRIPTION
**Pull request by Maxxen**
Continuation of #7403

This PR adds the possibility to log rows that failed to be parsed during ingestion when `ignore_errors=true` and the schema is specified into a separate "rejects table" through three new options to the `read_csv` table function
- `rejects_table=<string>` name of a temporary "rejects table" to create
- `rejects_limit=<int>` upper limit of rejects to record to the rejects table
- `rejects_recovery_columns=<varchar[]>` list of column names identifying additional columns whose values will be written to the rejects table for every failing row. Useful for scenarios where you have a "primary key" you know will be clean throughout the CSV that you want to be able to use to identify the cause of ingestion errors. If a recovery column would fail to parse we error out for real as if `ignore_errors=false`.

If the specified `rejects_table` already exists the query will fail, so you can't use the same rejects table multiple times in the same query, and will have to manage its lifetime yourself. However, rejects tables and their metadata are always created as temporary tables/stored in the object cache, and are thus local and bound to the lifetime of the connection.

Example from the tests:
Given the following `example.tsv`
```tsv
alice	10	1	20
bobby	12	2	NOT_A_NUMBER
charlie	7	3	30
```

```sql
SELECT * FROM read_csv(
    'test/sql/copy/csv/data/error/mismatch/example.tsv', 
    sep='\t',
    columns = {'name': 'VARCHAR', 'age': 'INTEGER', 'id': 'INTEGER', 'count': 'INTEGER'}, 
    rejects_table='csv_rejects_table',
    rejects_recovery_columns=['name', 'age'],
    ignore_errors=true
) ORDER BY ALL;
```

|  name   | age | id | count |
|---------|-----|----|-------|
| alice   | 10  | 1  | 20    |
| charlie | 7   | 3  | 30    |

```sql
FROM csv_rejects_table ORDER BY ALL;
```

| line | column | column_name | parsed_value |      recovery_columns      |                        error                         |                       file                        |
|------|--------|-------------|--------------|----------------------------|------------------------------------------------------|---------------------------------------------------|
| 2    | 3      | "count"     | NOT_A_NUMBER | {'name': bobby, 'age': 12} | Could not convert string 'NOT_A_NUMBER' to 'INTEGER' | test/sql/copy/csv/data/error/mismatch/example.tsv |

Note that this PR also fixes multiple bugs regarding line number calculations when multiple files/headers are present. This was always broken but used to not be a problem since we always just output the line of the first error, but now that we're interested in getting all the errors that is no longer the case.

### CODE REVIEW

**Comment by Maxxen** [comment]
@Mytherin [this CI run](https://github.com/duckdb/duckdb/actions/runs/5082185226/jobs/9138989497?pr=7681) seems like its running into an unrelated segfault (can't see how its affected by this PR), is this known or should I open a new issue?

**Comment by Mytherin** [comment]
Not a segfault but an assertion trigger :) but yes feel free to open an issue

**Review by pdet — CHANGES_REQUESTED** [review]
Hi Max, thanks for all the hard work on this. It looks very good, I've only added some small comments regarding the code.

Should we also add support for these options on the COPY statement and the read_csv functions from python?

Also, I think you just check for ignore_errors on type casting. Should we check for ignore errors on delimiters as well, IIRC this is done in the add_value function in the `base_csv_reader`


Could you add a couple more tests?
* Tests that trigger all exceptions you have created (e.g., wrong option setup, missing column from recovery_columns.
* Add tests with `READ_CSV_AUTO`
* Add tests that return more errors, at least over vector size.
* Test where there are more errors than the max number of errors we want to store.

File: src/execution/operator/persistent/base_csv_reader.cpp
Comment (Maxxen) [reply src/execution/operator/persistent/base_csv_reader.cpp:607]: This does create the names the same way as in the bind, I.e. `"column0"`, otherwise it uses the provided names (if there are any).

**Comment by Maxxen** [comment]
god dammint, I merged with master.

**Comment by Maxxen** [comment]
@Mytherin All green!

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great - some more minor comments:

File: src/execution/operator/persistent/csv_reader_options.cpp
Comment (Maxxen) [reply src/execution/operator/persistent/csv_reader_options.cpp:183]: Because the order matters, we can't handle the rejects_table option unless `columns` are resolved (either through argument or auto).

File: src/execution/operator/persistent/base_csv_reader.cpp
Comment (Maxxen) [reply src/execution/operator/persistent/base_csv_reader.cpp:610]: On second thought, this won't work, if we throw an error the rejects table doesn't get created. (because the transaction aborts?)

Comment (Mytherin) [reply src/execution/operator/persistent/base_csv_reader.cpp:610]: Hm, yes that is true. Perhaps a reason not to store the rejects table as an actual table?

Comment (Maxxen) [reply src/execution/operator/persistent/base_csv_reader.cpp:610]: I agree that it makes sense to throw an error and halt execution, but being able to specify the rejects table as a table was an explicit client request.

**Comment by Maxxen** [comment]
@Mytherin The regression failure is unrelated. This PR has been passing all checks for a while before I merged with feature this morning. Is there anything else blocking this from getting merged?

**Comment by Mytherin** [comment]
Thanks!

**Comment by whelanmike** [comment]
Thanks for this great feature. Is it possible to have the following scenario captured in the `rejects_table`. 
### Use Case:
Starting from the data-loading example above, I have added 1 more record that has an invalid 5th field.
This new record gets ignored silently and is not added to the `rejects_table`.

```csv
alice	10	1	20
bobby	12	2	NOT_A_NUMBER
charlie	7	3	30
NEW	0	1	2	3
```

**Pull request merged**
